fv/sockmap_test.go

```diff
@@ -0,0 +1,322 @@
+		strKeys = append(strKeys, strKey)
+	}
+	return strKeys
+}
+
+func testIPToHex(ip string) []string {
+	cidr := fmt.Sprintf("%s/32", ip)
+	hexen, err := bpf.CidrToHex(cidr)
+	Expect(err).NotTo(HaveOccurred())
+	prepend0xInPlace(hexen)
+	return hexen
+}
+
+func checkEndpointIP(felix *infrastructure.Felix, ip string) bool {
+	hexen := testIPToHex(ip)
+	log.WithFields(log.Fields{
+		"ip":    ip,
+		"hexen": hexen,
+	}).Info("Looking for IP.")
+	output, err := utils.Command(
+		"docker",
+		"exec",
+		felix.Container.Name,
+		"bpftool",
+		"--json",
+		"--pretty",
+		"map",
+		"dump",
+		"pinned",
+		"/sys/fs/bpf/calico/sockmap/calico_sk_endpoints_v1",
+	).CombinedOutput()
+	log.WithField("output", string(output)).Info("Dump of calico_sk_endpoints_v1")
+	Expect(err).NotTo(HaveOccurred())
+	var entries []mapEntry
+	Expect(json.Unmarshal(output, &entries)).NotTo(HaveOccurred())
+	log.WithField("entries", entries).Info("Checking map entry")
+	for _, entry := range entries {
+		if strSliceEqual(entry.Key, hexen) {
+			return true
+		}
+	}
+	return false
+}
+
+func unmarshalBpfToolSockhashDumpOutput(output []byte) []mapEntry {
+	var al []mapEntry
+	buf := bytes.NewBuffer(output)
+	// 0 - "key:", 1: actual key, 2: "value:", 3: actual value
+	phase := 0
+	for {
+		line, err := buf.ReadString('\n')
+		if err != nil {
+			return al
+		}
+		line = strings.TrimSpace(line)
+		if phase == 0 && strings.HasPrefix(line, "Found ") {
+			return al
+		}
+		if phase == 1 {
+			fields := strings.Fields(line)
+			prepend0xInPlace(fields)
+			entry := mapEntry{
+				Key: fields,
+			}
+			al = append(al, entry)
+		}
+		phase = (phase + 1) % 4
+	}
+}
+
+func getSockmapOpts() infrastructure.TopologyOptions {
+	opts := infrastructure.DefaultTopologyOptions()
+	opts.EnableIPv6 = false
+	opts.FelixLogSeverity = "debug"
+	opts.ExtraEnvVars["FELIX_XDPENABLED"] = "0"
+	opts.ExtraEnvVars["FELIX_SOCKMAPENABLED"] = "1"
+	return opts
+}
+
+var _ = infrastructure.DatastoreDescribe("with Felix using sockmap", []apiconfig.DatastoreType{apiconfig.EtcdV3 /*, apiconfig.Kubernetes*/}, func(getInfra infrastructure.InfraFactory) {
+	var (
+		infra   infrastructure.DatastoreInfra
+		felix   *infrastructure.Felix
+		host    *workload.Workload
+		ip      string
+		port    int
+		srcPort int
+	)
+
+	BeforeEach(func() {
+		if err := bpf.SupportsSockmap(); err != nil {
+			Skip(fmt.Sprintf("Sockmap acceleration not supported: %v", err))
+		}
+		infra = getInfra()
+		opts := getSockmapOpts()
+		felix, _ = infrastructure.StartSingleNodeTopology(opts, infra)
+		ip = "10.65.0.2"
+		port = 8055
+		srcPort = 8056
+		host = workload.Run(
+			felix,
+			"service",
+			"default",
+			ip,
+			fmt.Sprintf("%d", port),
+			"tcp")
+		host.ConfigureInDatastore(infra)
+	})
+
+	AfterEach(func() {
+		host.Stop()
+		felix.Stop()
+		func() {
+			opts := getSockmapOpts()
+			felix, _ = infrastructure.StartSingleNodeTopology(opts, infra)
+			defer felix.Stop()
+			outputBytes, err := utils.Command(
```

Comment:
Think you can do `felix.ExecOutput()`

Reply:
Won't felix already be stopped/stopping due  to the call above?  Should this run before we stop the felix container?

Reply:
We stop felix and then start a new one to clean up the old sockmap state completely. The new sockmap state will be empty, so we can easily detach the sockhash map from the cgroup manually without worrying about some weird corner cases resulting from failed test. I'll document it.

Used the `ExecOutput` function too.